core/src/memory/host_memory_resource.cpp

```diff
@@ -12,21 +12,60 @@
 #include "vecmem/utils/debug.hpp"
 
 // System include(s).
+#include <algorithm>
+#include <cassert>
 #include <cstdlib>
+#include <memory>
 
 namespace vecmem {
 
-void *host_memory_resource::do_allocate(std::size_t bytes, std::size_t) {
+void *host_memory_resource::do_allocate(std::size_t bytes,
+                                        std::size_t alignment) {
 
-    void *ptr = std::malloc(bytes);
-    VECMEM_DEBUG_MSG(4, "Allocated %lu bytes at %p", bytes, ptr);
+#ifdef VECMEM_HAVE_STD_ALIGNED_ALLOC
+    // Rely on std::aligned_alloc to give us properly aligned memory.
+    // Note that std::aligned_alloc has a number of quirks about the parameters
+    // that it would accept. Hence the tweaking of the alignment and size
+    // values.
+    const std::size_t align = std::max(alignment, alignof(std::max_align_t));
+    const std::size_t size = bytes + (align - (bytes % align));
+    void *ptr = std::aligned_alloc(align, size);
+#else
+    // Ask for enough memory that will allow us to be able to align the
+    // the returned pointer for sure, **and** be able to store the
+    // pointer to the beginning of the unaligned memory blob before
+    // the aligned pointer.
+    std::size_t alloc_bytes = bytes + alignment + sizeof(void *) - 1;
+    void *unaligned_ptr = std::malloc(alloc_bytes);
+    // Get an aligned pointer that has space "before it" for the
+    // pointer to the beginning of the unaligned space.
+    void *ptr = static_cast<char *>(unaligned_ptr) + sizeof(void *);
+    std::align(alignment, bytes, ptr, alloc_bytes);
```

Comment:
Note that the size `alloc_bytes` has shrunk by the size of a void pointer here, which we should account for so that `std::align` can properly sanity check the call.